[CNSMR-2107] Create UnitTestingViewModels.md (#176)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/UnitTestingViewModels.md

## Unit testing view models

### What do we test in the view model unit tests?

- view model state changes
- routing events

> - analytics events

Reviewer: Should they be in a different file/test? IMO it makes it easier to read/scan.

Author: this can lead to unnecessary duplicated tests, as the same action can result in multiple events (state change, routing, analytics event) happening all together

### Creating a new test case

> To create a test case for a view model use an Xcode template called `BentoViewModelTests` and provide the view model name (without `ViewModel` suffix). This template will create a basic structure for the test case that you will need to adjust to your specific implementation.

Reviewer: ```suggestion
To create a test case for a view model use the Xcode template `BentoViewModelTests` and provide the view model name (without `ViewModel` suffix). This template will create a basic structure for the test case that you will need to adjust to your specific implementation.

E.g: your feature is called `MyFeature`, and is has a `MyFeatureViewModel`. You just need to name the file `MyFeature`, the template will automatically add the proper suffix making it `MyFeatureViewModelTests`.
```

Reviewer 2: @diegopetrucci 
> ... and is has ... 

... and *it* has ... would you mind if I edit your suggestion?

Reviewer 2: also before the "and" no comma is needed.

Reviewer: Sure big man, feel free

Reviewer 2: just to be clear: I'll do it just when it's a typo in a suggestion. Otherwise it's a rabbit hole. 😉

### Writing a test

> When implementing a test use `perform(stub:when:assert:)` herlper method from `BabylonTestUtilities`.

Reviewer 3: ```suggestion
When implementing a test use `perform(stub:when:assert:)` helper method from `BabylonTestUtilities`.
```

Reviewer 3: Maybe you could provide the full signature here? Because not having it memorised it's quite hard to follow (and this is aimed at new starters mostly, right?).

I think the reader should be able to have a more or less complete understanding of what is happening without having to look at the code in the project, and this doc only has truncated signatures and a non-structured description of parameters. I believe providing a complete, if short, example first, with full signatures so that the reader understands what's going on, and then dissecting it in following sections would be more conductive to learning.

Author: do you thin `perform<ViewModel: BoxViewModel>(stub: (TestScheduler) -> ViewModel, when: (ViewModel, TestScheduler) -> Void, assert: ([ViewModel.State]) -> Void)` will make it simpler to understand? I have doubts 😄

> When implementing a test use `perform(stub:when:assert:)` herlper method from `BabylonTestUtilities`.

Reviewer: ```suggestion
When implementing a test use the `perform(stub:when:assert:)` helper method from `BabylonTestUtilities`.
```

> - `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.

Reviewer 3: ```suggestion
- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events dispatch via its `advance` method later in the `when` closure.
```

> - `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.

Reviewer: ```suggestion
- the `stub` parameter is used to create a view model and to provide it its stubbed dependencies via the `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, which allows to control signals events dispatch via its `.advance` method (to be used in later moment in the `when` closure).
```

> - `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 

Reviewer 3: ```suggestion
- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call its `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
```

> - `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 

Reviewer: ```suggestion
- the `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call its `.send` method and provide it with an action. This effectively simulates user interaction with a screen managed by this view model. 
```

> - `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 

Reviewer 2: All other sections have code example. Could we have here as well? It's very helpful. I just fully understand seeing an example. 😬

> After sending an event you will need to call `scheduler.advance()`. Until this method is called at least once no events will be produced by any signal in the view model state machine. As soon as `advance` is called it will "release" the first event and so on. 

Reviewer: ```suggestion
After sending an event you will need to call `scheduler.advance()`. Until this method is called at least once no events will be produced by any signal in the view model state machine. As soon as `advance` is called it will "release"/trigger the first event and so on. 
```

> It is also possible to advance scheduler by specific time interval with `advance(by:)` method instead of using `sleep`. This is useful when any signal in the view model state machine uses timer (i.e. with `delay` operator)

Reviewer: ```suggestion
It is also possible to advance the scheduler by a specific time interval with `advance(by:)` method instead of using `sleep`. This is useful when any signal in the view model state machine uses a timer (i.e. with `delay` operator)
```

Example:

> 

Reviewer 4: Maybe we could have an example of `stub` too?

Author: essentially it should be just calling `makeViewMode` with some parameters, and that is already in the template, that's why I didn't put it there 🤷‍♂

Reviewer 4: 👍

E.g: your feature is called `MyFeature` and it has a `MyFeatureViewModel`. You just need to name the file `MyFeature`, the template will automatically add the proper suffix making it `MyFeatureViewModelTests`.

### Writing a test

When implementing a test use the `perform(stub:when:assert:)` helper method from `BabylonTestUtilities`.

Example:

```swift
perform(
    stub: { scheduler in
        self.makeViewModel(scheduler: scheduler)
    }
    when: { viewModel, scheduler in
        viewModel.send(action: ...)
        scheduler.advance()
        viewModel.send(action: ...)
        scheduler.advance()
    },
    assert: { states in
        expect(states).to(equal([
            makeState(status: .initial, intent: nil),
            makeState(status: .loading, intent: nil),
            ...  
        ]))
    
        expect(analytics.accumulatedEvents).to(equal([
            AnyEquatable(Tracking.ActionEvents.event),
        ]))
    }
)
```

> - `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via it's `advance` method later in the `when` closure.

Reviewer 3: ```suggestion
- `stub` parameter is used to create a view model and provide it its stubbed dependencies via `makeViewModel` method. It accepts a `scheduler` parameter that you must pass to the view model constructor. This scheduler is an instance of a `TestScheduler`, it allows to control signals events displatch via its `advance` method later in the `when` closure.
```

> - `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call it's `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 

Reviewer 3: ```suggestion
- `when` closure is where the actual interaction with view model should happen. To "interact" with a view model you should call its `send` method and provide an action. This effectively simulates user interaction with a screen managed by this view model. 
```
